internal/devpkg/narinfo_cache.go

```diff
@@ -47,41 +48,61 @@ func (p *Package) IsInBinaryCache() (bool, error) {
 
 	// Check if the narinfo is present in the binary cache
 	isNarInfoInCache.lock.RLock()
-	exists, ok := isNarInfoInCache.status[p.Raw]
+	status, statusExists := isNarInfoInCache.status[p.Raw]
 	isNarInfoInCache.lock.RUnlock()
-	if !ok {
-		return false, errors.Errorf(
-			"narInfo cache miss: %v. Call FillNarInfoCache before invoking IsInBinaryCache",
-			p.Raw,
-		)
+	if !statusExists {
+		// Fallback to synchronously filling the nar info cache
+		if err := p.fillNarInfoCache(); err != nil {
```

Comment:
For potentially simpler code. By definition, `fillNarInfoCache()` only returns a nil error if it succeeded. So you could do:

```golang
err := p.fillNarInfoCache()
return err == nil, err
```

Additionally, if `fillNarInfoCache` does

```golang
if isNarInfoInCache.status[p.Raw] {
  return nil
}
```

(no need to lock because we never remove from cache)

Then this entire function could be:

```golang
func IsInBinaryCache()
  if eligible, err := p.isEligibleForBinaryCache(); err != nil {
    return false, err
  } else if !eligible {
    return false, nil
  }
  err := p.fillNarInfoCache()
  return err == nil, err
}
```

(maybe rename `fillNarInfoCache` to `fillNarInfoCacheIfNeeded` for clarity)

Reply:
Nope, this doesn't quite work for two reasons:

1. fillNarInfoCache stores the _boolean_ of whether the value is stored or not. A _false_ value for a key in `isNarInfoInCache` does not indicate an error.

2. Golang maps are not safe for concurrent reads and writes, so we will need a lock.

Reply:
@savil you are correct on (1) but not on (2). 

For (2), reading a `true` is only unsafe if we remove and/or change values. Since we should never remove/change values within a single command, I think this is safe and improves code readability quite a bit.

For (1), we can modify `fillNarInfoCache()` to return `(status, error)`. The early return continues to be safe because if something is set to true it never changes.

An additional benefit of this approach is that we limit cache access to a single function.

Reply:
clarified why (2) is correct in https://github.com/jetpack-io/devbox/pull/1473#discussion_r1323495152